Multi-Scale Tilt Depth Estimation

mtb-za opened the pull request:
This is an early PR for getting some new filters in.

Given that this is something fairly new to me, I am making it so early from my side so that other people can hopefully pull me up short if I start doing something particularly stupid.

So far, I have implemented the Tilt Angle, but am not sure quite what the output needs to look like. I have modified another of the cookbook recipes to test it, but if someone else could just take a look and let me know, it would be appreciated.

**UPDATE 25-05-2015 by @leouieda**: Fixes #203 

- [ ] Make tests for new code
- [ ] Create/update docstrings
- [ ] Include relevant equations and citations in docstrings
- [ ] Code follows PEP8 style conventions
- [ ] Code and docs have been spellchecked
- [ ] Include new dependencies in docs, requirements.txt, README, and .travis.yml
- [ ] Documentation builds properly
- [ ] All tests pass
- [ ] Can be merged
- [ ] Changelog entry (leave for last)

leouieda commented on fatiando/gravmag/transform.py:
```diff
@@ -112,6 +114,59 @@ def tga(x, y, data, shape):
+        The tilt angle of the total field.
+
+    """
+    horiz_deriv = derivx(x, y, data, shape)/derivy(x, y, data, shape)
```
@mtb-za from the Miller and Singh paper, the horizontal derivative should be df/dh = sqrt(df/dx^2 + df/dy^2)

leouieda commented on fatiando/gravmag/transform.py:
```diff
@@ -112,6 +114,59 @@ def tga(x, y, data, shape):
+
+    """
+    horiz_deriv = derivx(x, y, data, shape)/derivy(x, y, data, shape)
+    tilt = numpy.arctan( ( derivz(x, y, data, shape) ) / numpy.sqrt( horiz_deriv**2 ) )
```
Computing the arctan of a division can quickly get unstable for small values of the dividend. Fortunately, there is a [numpy.arctan2](http://docs.scipy.org/doc/numpy/reference/generated/numpy.arctan2.html) that computes this in a stable way, even for dividend = 0 (this functions exists in every math library I know).

mtb-za replied:
What should I be using as the second parameter to (numpy.arctan)[http://docs.scipy.org/doc/numpy/reference/generated/numpy.arctan2.html]? Just using the calculated tilt value again seems to be working, but I am not sure that this is entirely valid.

leouieda replied:
`arctan2(x, y)` is equivalent to `arctan(x/y)` but is more stable and doesn't fail if `y == 0`. So in practice you never compute the arctan of a division using `arctan(x/y)`. You use `arctan2` instead.

leouieda commented on fatiando/gravmag/transform.py:
```diff
@@ -112,6 +114,59 @@ def tga(x, y, data, shape):
+    Van Buren, Reece. 2013. “Multi-Scale Tilt Depth Estimation.” MSc Thesis,
+    Johannesberg: University of the Witwatersrand.
+    URI: http://mobile.wiredspace.wits.ac.za/bitstream/handle/10539/14011/MSc_Dissertation_RvanBuren_2013.pdf.
+    """
```
When leaving functions empty to be filled later, put a `pass` inside it. Missing that will cause an import error for the module and the tests will break.

leouieda commented:
@mtb-za thanks for adding a new processing function! There are so many for potential fields and I couldn't possibly implement them all.

leouieda commented on fatiando/gravmag/transform.py:
```diff
@@ -112,6 +114,59 @@ def tga(x, y, data, shape):
+        strange units! I strongly recommend converting the data to SI
+        **before** calculating the derivative (use one of the unit conversion
+        functions of :mod:`fatiando.utils`). This way the derivative will be in
+        SI units and can be easily converted to what unit you want.
```
It doesn't really matter what the units of the data are because the tilt angle will be in radians anyway, right? The ratio df/dz / df/dh will be unitless.

mtb-za commented:
Yeah, I am busy trying to get my head around all of these things still.

I have addressed most of your comments, I think.

Hopefully I can get the tilt done soon, while I try and get my head around the way MSTDE really works.

leouieda commented:
@mtb-za you can push your changes to this branch as soon as you make them. That way the PR is updated. It's always best to make  many small commits than a really big one involving many files. They are easier to revert if needed.

leouieda commented:
I've been thinking of a way to test the results of the tilt. It basically calculates the vertical angle of the gradient of some data. One way to test it would be make some fake data using an inclined plane. You could generate the plane using a known angle and see if the tilt computes that angle correctly on the entire grid.

Another test case could be a grid with constant value. The tilt should also be 90º because the horizontal gradient will be zero.

mtb-za commented:
Hmm.... I seem to have messed up this merge. Ah well. Will try and fix it tomorrow.

leouieda commented:
Seems like you left the git merge markers in the file. You can revert the merge commit (see [this stackoverflow question](http://stackoverflow.com/questions/2389361/undo-a-git-merge)). After that you'll need to force a push to your github repo with `git push -f`.

If you need any help resolving the merge conflicts, let me know. Github has a [useful guide](https://help.github.com/articles/resolving-a-merge-conflict-from-the-command-line/) for resolving conflicts.

leouieda commented on fatiando/gravmag/transform.py:
```diff
@@ -299,3 +397,31 @@ def _fftfreqs(x, y, shape, padshape):
+
+    A useful thing used in many tilt angle filters:
+
+    sqrt( \\frac{df}{dx}^2 + \\frac{df}{dy}^2 )
```
You need to put the line above in a `.. math::` block for the it to render in the docs [like this](https://github.com/fatiando/fatiando/blob/master/fatiando/gravmag/transform.py#L90).

Also, if you declare the docstring as a "raw string" then you don't have to write `\\` every time. Do this by opening the string with `r"""`.

leouieda commented on fatiando/gravmag/transform.py:
```diff
@@ -118,6 +125,97 @@ def tga(x, y, data, shape, method='fd'):
+    horiz_deriv = thdr(x, y, data, shape)
+    vert_deriv = derivz(x, y, data, shape)
+    tilt_value = vert_deriv/horiz_deriv
+    tilt = numpy.arctan2( tilt_value, tilt_value )
```
@mtb-za no need to compute `tilt_value`. You can call `arctan2(vert_deriv, horiz_deriv)`. That's what `arctan2` is for: avoiding the division. If `horiz_deriv` is zero (or close to zero), the division will break `arctan2` will still give you 90º.

leouieda commented on fatiando/gravmag/transform.py:
```diff
@@ -118,6 +125,97 @@ def tga(x, y, data, shape, method='fd'):
+
+    **References**
+
+    Van Buren, Reece. 2013. “Multi-Scale Tilt Depth Estimation.” MSc Thesis,
```
These two " characters are causing errors. They are no ASCII characters (probably from copying from a PDF file). You should replace them with the proper " from your keyboard or just remove them.

leouieda commented on fatiando/gravmag/transform.py:
```diff
@@ -118,6 +125,97 @@ def tga(x, y, data, shape, method='fd'):
+
+    * tilt : 1D-array
+        The tilt angle of the total field.
+
```
Could you add the reference here for Miller and Singh (1994) as you did above?

mtb-za commented on fatiando/gravmag/transform.py:
```diff
@@ -118,6 +125,106 @@ def tga(x, y, data, shape, method='fd'):
+    MSc_Dissertation_RvanBuren_2013.pdf.
+    """
+    RTP_TDR = tilt(x, y, data, shape)
+    AREA = numpy.arctan2( numpy.tan( (numpy.pi/2)-numpy.arctan2(RTP_TDR) ) )
```
I am not really sure yet how to approach this sensibly using arctan2. Essentially, I need to sort out what the second parameter needs to be for the two calls to arctan2 in this line.

leouieda replied:
Has this been published yet? If not, could you point me to the page/equation in the thesis that you're implementing here?

leouieda commented on fatiando/gravmag/transform.py:
```diff
@@ -118,6 +125,106 @@ def tga(x, y, data, shape, method='fd'):
+
+    Miller, Hugh G, and Vijay Singh. 1994. "Potential Field Tilt --- a New
+    Concept for Location of Potential Field Sources."
+    Journal of Applied Geophysics 32 (2--3): 213–17.
```
@mtb-za keep an eye out for non-ASCII characters in your code because Python can be a bit annoying about it. For example, the dash in "213-17" is not really an ASCII "-". Notice that it's slightly longer. This happens all the time when copy/pasting a reference.

leouieda commented:
Hi @mtb-za, any new progress on this?

mtb-za commented:
I thought I had replied to your last question a few days ago already, sorry.

Anyway, I was in the field for a while, but am back and working on stuff again. I will try and let you know in future.

I am pretty happy that I have implemented the TDX (Horizontal Tilt Angle) correctly, so am now looking at what else I need to do to get MSTDE in place. Busy dissecting the MATLAB code in the thesis more closely over the next couple of days to see exactly what is going on and how to convert it to how things are happening within Fatiando.

leouieda commented:
No problem, I was just checking up on the currently open PRs. 
I say from your blog that you were out (nice posts by the way).

Would you consider splitting this into multiple PRs? You mentioned that the TDX is implemented already. The reason I ask is that long PRs tend to get very out of sync with `master` (I have a couple that I should have split long ago, but you know, time). It's easier to separate the PR into small  things that can be merged independently and quickly. That way it's less work for  you to keep up with the changes in the code base. And we're having a lot of backwards breaking changes lately. 

If you agree, then I think it would be best to leave only the TDX code in this branch.  We can then finalize this (tests + docs) and get it merged. After that you can open another PR for the next function you need.

leouieda commented:
I'm closing this Pull Request because it's been inactive for a long time. The branch is probably very out of sync with `master`.

If you want to continue working of this, please [sync your fork](https://help.github.com/articles/syncing-a-fork/), update your branch, and re-submit the PR.  Let me know if you need any help doing this!

Sorry for the hassle! This is not a critique of your work, just a way to keep things fresh in the project. I hope you'll consider re-submitting.

**This is a message template.**
